tests/net/adversary.rs

```diff
@@ -7,14 +7,291 @@
+//!
+//! The following capabilities are explicitly not included:
+//!
+//! 1. Dropping of messages. The networking layer is expected to ensure that no messages
+//!    are lost. A node that drops messages regardless is considered faulty in real-world
+//!    deployments.
+//! 1. Forging message senders. The networking layer is also expected to sign messages and ensure
+//!    that they are not forged.
+//!
+//! ## Handles
+//!
+//! The adversary manipulates the network and nodes exclusively through handles that ensure they do
+//! not violate the constraints defined above. Handles are either mutable or immutable and can, in
+//! some cases be upgraded to actual references, if the underlying node is faulty (see
+//! `NodeHandle::node()` and `NodeHandle::node_mut()`).
 
-use net::{CrankError, NetMessage, VirtualNet};
+use std::cmp;
 
 use hbbft::messaging::{DistAlgorithm, Step};
+use net::{CrankError, NetMessage, Node, VirtualNet};
+
+/// Immutable network handle.
+///
+/// Allows querying public information of the network or getting immutable handles to any node.
+pub struct NetHandle<'a, D: 'a>(&'a VirtualNet<D>)
+where
+    D: DistAlgorithm;
+
+impl<'a, D: 'a> NetHandle<'a, D>
+where
+    D: DistAlgorithm,
+{
+    /// Returns a node handle iterator over all nodes in the network.
+    #[inline]
+    pub fn nodes(&self) -> impl Iterator<Item = NodeHandle<D>> {
+        self.0.nodes().map(NodeHandle::new)
+    }
+
+    /// Returns an iterator over all faulty in the network.
+    ///
+    /// Instead of a handle, returns the node directly, as the adversary gets full access to all
+    /// nodes in the network.
+    #[inline]
+    pub fn faulty_nodes(&self) -> impl Iterator<Item = &Node<D>> {
+        // FIXME: Add an API to handle a step?
+        // Not wrapped in a `NodeHandle`, the adversary gets full access to their own nodes.
+        self.0.faulty_nodes()
+    }
+
+    /// Returns a node handle iterator over all correct nodes in the network.
+    #[inline]
+    pub fn correct_nodes(&self) -> impl Iterator<Item = NodeHandle<D>> {
+        self.0.correct_nodes().map(NodeHandle::new)
+    }
+
+    /// Returns an iterator over all messages in the network.
+    #[inline]
+    pub fn messages(&'a self) -> impl Iterator<Item = &'a NetMessage<D>> {
+        self.0.messages()
+    }
+
+    /// Returns a handle to a specific node handle.
+    #[inline]
+    pub fn get(&self, id: D::NodeId) -> Option<NodeHandle<D>> {
+        self.0.get(id).map(NodeHandle::new)
+    }
+}
+
+/// Queue position.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum QueuePosition {
+    /// Front of the queue (equivalent to `Before(0)`).
+    Front,
+    /// Back of the queue.
+    Back,
+    /// Before a specific a position.
+    Before(usize),
+}
+
+/// Mutable network handle.
+///
+/// Allows reordering of messages, injecting new ones into the network queue and getting mutable
+/// handles to nodes.
+pub struct NetMutHandle<'a, D: 'a>(&'a mut VirtualNet<D>)
```

Comment:
Is there a reason why you used `where` here instead of keeping everything on a single line?

```rust
pub struct NetMutHandle<'a, D: 'a, D: DistAlgorithm>(&'a mut VirtualNet<D>);
```